### ISSUE TITLE
Issues with leader election when peers go down

### ISSUE THREAD

**Issue by ccampbell** [ipfs/ipfs-cluster#160]
I have been playing around with ipfs-cluster to try to set up a dynamic cluster. I have read through the docs and what available information I can find on other tickets. What I am noticing is that the cluster only seems to be working and healthy if the setup happens like this (assume three nodes A, B, C - using `leave_on_shutdown` set to `true`).

1. Start up ipfs service on node A with cluster secret and no specified peers (A becomes the leader)
1. Start up ipfs service on node B with cluster secret and --bootstrap set to node A (B joins the cluster as a new peer)
1. Start up ipfs service on node C with cluster secret and --bootstrap set to node A (C joins the cluster as a new peer)

So that is all fine and works great. The issue has to do with what happens when peers have issues. If I kill ipfs-cluster-service on node B it correctly is removed and then when it starts up (again bootstrapping to A) it comes back which is expected. 

Now If I kill the ipfs-cluster-service on node A (the leader) that is where there is a problem. Since peer A has been elected the leader, when it goes down there is no longer a leader present.  I would expect either peer B or C to become the leader in this case, but I could not find anything specific in the documentation to say what the expected behavior here is. They start logging about how there is no leader present, but neither one takes over as the leader. The second issue is that as soon as A comes back (bootstrapping to either B or C) it does not become the leader again according to all peers. Once the cluster gets into this state, the only way to get things back to normal is to kill the service on peers B and C and then restart them bootstrapping to A. You can understand why this is not an ideal situation for a peer to peer network.

Another issue which seems like it might be related is that I have noticed when I bootstrap a node D (running on my Mac - the other 3 are running on Linux), only nodes A and C pick up the new peer. I  can manually add it using the peer add command on node B, and it will pick it up, but when it goes down it is only automatically removed from A and C, so B never picks up on the changes. I am not sure if that actually leads to any problems or not.

It seems like it could be some kind of race condition. I have not had a chance to dig through the code and since this is kind of a complicated issue to reproduce, I don’t have code to reproduce it right now, but I am hoping you can provide some guidance or idea about why this is happening, what might be causing it, and what I might be able to do to get around it. Not sure if it is related at all, but node A is running in Amsterdam, B in New Jersey, C in Tokyo.

Also I am curious what the expected behavior is in this situation since I couldn’t find it exactly outlined in the docs. I am happy to help debug this further. Thanks so much.

**Comment by hsanjuan** [comment]
Hi!

I am going to assume that you are soft-killing your nodes, so they have time to perform the `leave_on_shutdown` procedures are remove themselves from the cluster.

In an A, B, C cluster, removing the leader (A) from it will leave a 2-peer cluster and trigger an election. I think might be then dead-locked as none of the candidates nodes (B,C) obtain a majority of votes to proclaim themselves leaders. I am not sure if you plan to remove peers more or less permanently, but if it is just simulating an unexpected downtime, or a reboot, you would rather set `leave_on_shutdown` to false: this make sure that the cluster size remains at 3, and saves problems from updating the cluster's topology (https://github.com/ipfs/ipfs-cluster/blob/master/docs/ipfs-cluster-guide.md#dynamic-cluster-membership-considerations)

Regarding adding a D node, make sure you are in a healthy cluster first (I am not sure it was the case). If B did not pick it up, it would mean that B is virtually split from the A-C cluster.

Last of all, there is a real possibility that Raft latency/timeout options are not tuned for multi-region/high latency operation. As pointed out in #157, this is something we aim to address in the short term, as it is clearly a pain point for some users.

**Comment by ccampbell** [comment]
Thanks for your reply. I am testing using `SIGTERM` for the killing. I am going to play around with trying `leave_on_shutdown` set to `false` for the leader to see if that makes a difference.

I am noticing very strange behavior, and I can’t figure out a pattern. Sometimes just the process of bootstrapping  a new node to the cluster (node D) will cause some of the existing nodes to lose track of the leader and start logging 

```
error broadcasting metric: leader unknown or not existing yet
```

After that pinning to the cluster will fail which is expected if it doesn’t think it has enough peers. I have noticed that sometimes when that happens, killing the service on the node that had just joined and then bootstrapping it again can trigger a re-election. Also killing the service on the node that is logging errors and starting that one again can trigger it as well.

I am definitely still seeing where when a specific node is removed some other peers will remove it and some will not, but still fail to connect to it and `peers ls` will still list it with:

```
Qmdr4GTYHCa7BGWE4jtVUpqwN6MKWZWiHZj618xDDYo6Ys | ERROR: dial backoff
```

---

I am just trying to simulate the condition of nodes being added and removed. The use case here will eventually be a peer to peer network where anyone can add their own peer to offer up storage, so it has to be very fault tolerant, and it worries me that sometimes just the process of bootstrapping a new node to the cluster can cause the whole cluster to end up in an unhealthy state in my testing.

**Comment by hsanjuan** [comment]
Would be super useful if you pin down a way to reproduce it reliably. Alternatively, you are saying that bootstrapping a 4th node to an otherwise working cluster fails [sometimes]?

**Comment by hsanjuan** [comment]
Also @ccampbell , maybe you can send `ipfs-cluster-ctl peers ls` for each A,B,C,D peer?

**Comment by hsanjuan** [comment]
@ccampbell, just a quick word that I'm working on #131 and I will attempt to reproduce your problem with latest `0.2.0` and with the new raft this week. Hopefully I can get a clearer picture of what's failing.

**Comment by hsanjuan** [comment]
@ccampbell I have been doing some testing with the last version of master today.

I have discovered and fixed few issues with saving and reading configuration.

I can have 3 -A,B,C- nodes with `leave_on_shutdown` set to `true`. Killing the leader causes it to leave cleanly. Bootstrapping the node to the others works fine too.

I think you might have been affected by a misconfiguration. We made recently made a change that saves all existing cluster peers as `bootstrap` peers when leaving the cluster and empties the `cluster_peers`. Otherwise, re-starting the peer would ignore bootstrap and try to use the former list of cluster peers without letting everyone know that it wishes to re-join.

This was non-obvious behaviour. A peer that has left the cluster should only rejoin with `bootstrap`. It should not be started on its own (as a single peer cluster) as it's state should not be allowed to diverge. If that is the case, it will be necessary to fully remove the `~/.ipfs-cluster/ipfs-cluster-data` folder before so it can simply download the full state from other peers.

I will be publishing version `0.2.2` soon (`0.2.0` and 0.2.1` have reading/writing configuration bugs). In the meantime, I think using master should probably solve the issues you were having.

Also, since your nodes are very far away, make sure you tune your `consensus.raft` options in the configuration and increase the timeouts.

**Comment by ccampbell** [comment]
Thanks so much! I will test it out again using master and let you know what I find

**Comment by ccampbell** [comment]
Okay now I am seeing some new issues. For one when I start up a single node A it fails to elect itself as the leader and immediately shuts down after some period of time. The output is:

```
16:01:13.454  INFO    cluster: IPFS Cluster v0.2.1 listening on: cluster.go:90
16:01:13.455  INFO    cluster:         /ip4/127.0.0.1/tcp/9096/ipfs/QmVDrbRYZckmBaNYexBw8Kwkhz9fPKGgccN5AARkPSKmrz cluster.go:92
16:01:13.455  INFO    cluster:         /ip4/127.0.0.2/tcp/9096/ipfs/QmVDrbRYZckmBaNYexBw8Kwkhz9fPKGgccN5AARkPSKmrz cluster.go:92
16:01:13.455  INFO    cluster:         /ip4/80.209.230.17/tcp/9096/ipfs/QmVDrbRYZckmBaNYexBw8Kwkhz9fPKGgccN5AARkPSKmrz cluster.go:92
16:01:13.456  INFO  consensus: starting Consensus and waiting for a leader... consensus.go:68
16:01:13.459  INFO    restapi: REST API: /ip4/127.0.0.1/tcp/9094 restapi.go:266
16:01:13.459  INFO   ipfshttp: IPFS Proxy: /ip4/127.0.0.1/tcp/9095 -> /ip4/127.0.0.1/tcp/5001 ipfshttp.go:158
16:01:43.459 ERROR    cluster: consensus start timed out cluster.go:361
16:01:43.460  INFO    cluster: shutting down IPFS Cluster cluster.go:415
16:01:43.460 WARNI    cluster: Attempting to leave Cluster. This may take some seconds cluster.go:419
16:01:43.490  INFO  consensus: Raft Leader elected: QmVDrbRYZckmBaNYexBw8Kwkhz9fPKGgccN5AARkPSKmrz raft.go:132
16:01:43.492  INFO    cluster: removing Cluster peer QmVDrbRYZckmBaNYexBw8Kwkhz9fPKGgccN5AARkPSKmrz peer_manager.go:55
16:01:43.494  INFO  consensus: peer removed from global state: <peer.ID VDrbRY> consensus.go:385
16:01:45.494  INFO  consensus: stopping Consensus component consensus.go:161
16:01:45.496  INFO     config: Saving configuration config.go:289
16:01:45.496  INFO    monitor: stopping Monitor peer_monitor.go:161
16:01:45.497  INFO    restapi: stopping Cluster API restapi.go:284
16:01:45.497  INFO   ipfshttp: stopping IPFS Proxy ipfshttp.go:532
16:01:45.497  INFO pintracker: stopping MapPinTracker maptracker.go:116
```

And the process kills itself.

I have increased the raft heartbeat timeout and election timeout to 10s.

When I start up another node bootstrapping to the first while the first is booting up, the A node returns:

```
16:04:48.614  INFO    cluster: new Cluster peer /ip4/80.209.230.16/tcp/9096/ipfs/QmYVdQTKmbyz1nU4bR9wCzfqLfmYpnV7zjyPdnno5c8QZF peer_manager.go:41
16:04:59.938  INFO  consensus: Raft Leader elected: QmVDrbRYZckmBaNYexBw8Kwkhz9fPKGgccN5AARkPSKmrz raft.go:132
16:05:14.184 ERROR    cluster: consensus start timed out cluster.go:361
16:05:14.184  INFO    cluster: shutting down IPFS Cluster cluster.go:415
16:05:14.184 WARNI    cluster: Attempting to leave Cluster. This may take some seconds cluster.go:419
16:05:14.186  INFO  consensus: Raft Leader elected: QmVDrbRYZckmBaNYexBw8Kwkhz9fPKGgccN5AARkPSKmrz raft.go:132
16:05:14.186  INFO  consensus: Raft Leader elected: QmVDrbRYZckmBaNYexBw8Kwkhz9fPKGgccN5AARkPSKmrz raft.go:132
16:05:14.187  INFO    cluster: removing Cluster peer QmVDrbRYZckmBaNYexBw8Kwkhz9fPKGgccN5AARkPSKmrz peer_manager.go:55
16:05:14.189 ERROR       raft: Failed to AppendEntries to QmYVdQTKmbyz1nU4bR9wCzfqLfmYpnV7zjyPdnno5c8QZF: operation not supported with current protocol version logging.go:28
16:05:14.201 ERROR       raft: Failed to AppendEntries to QmYVdQTKmbyz1nU4bR9wCzfqLfmYpnV7zjyPdnno5c8QZF: operation not supported with current protocol version logging.go:28
16:05:14.214 ERROR       raft: Failed to AppendEntries to QmYVdQTKmbyz1nU4bR9wCzfqLfmYpnV7zjyPdnno5c8QZF: operation not supported with current protocol version logging.go:28
16:05:14.308 ERROR       raft: Failed to AppendEntries to QmYVdQTKmbyz1nU4bR9wCzfqLfmYpnV7zjyPdnno5c8QZF: operation not supported with current protocol version logging.go:28
16:05:14.447 ERROR       raft: Failed to AppendEntries to QmYVdQTKmbyz1nU4bR9wCzfqLfmYpnV7zjyPdnno5c8QZF: operation not supported with current protocol version logging.go:28
16:05:14.595 ERROR       raft: Failed to AppendEntries to QmYVdQTKmbyz1nU4bR9wCzfqLfmYpnV7zjyPdnno5c8QZF: operation not supported with current protocol version logging.go:28
16:05:14.696 ERROR  consensus: raft cannot remove peer: leadership lost while committing log raft.go:250
16:05:14.697 ERROR  consensus: raft cannot add peer: leadership lost while committing log raft.go:233
16:05:14.830 ERROR       raft: Failed to AppendEntries to QmYVdQTKmbyz1nU4bR9wCzfqLfmYpnV7zjyPdnno5c8QZF: operation not supported with current protocol version logging.go:28
16:05:29.697 ERROR  p2p-gorpc: context deadline exceeded client.go:125
16:05:29.697 ERROR  p2p-gorpc: context deadline exceeded client.go:125
16:05:29.897 ERROR  p2p-gorpc: context deadline exceeded client.go:125
16:05:32.147 ERROR       raft: Failed to make RequestVote RPC to QmYVdQTKmbyz1nU4bR9wCzfqLfmYpnV7zjyPdnno5c8QZF: operation not supported with current protocol version logging.go:28
```

The B node returns

```
16:04:48.502  INFO    cluster: IPFS Cluster v0.2.1 listening on: cluster.go:92
16:04:48.502  INFO    cluster:         /ip4/127.0.0.1/tcp/9096/ipfs/QmYVdQTKmbyz1nU4bR9wCzfqLfmYpnV7zjyPdnno5c8QZF cluster.go:94
16:04:48.502  INFO    cluster:         /ip4/127.0.0.2/tcp/9096/ipfs/QmYVdQTKmbyz1nU4bR9wCzfqLfmYpnV7zjyPdnno5c8QZF cluster.go:94
16:04:48.502  INFO    cluster:         /ip4/80.209.230.16/tcp/9096/ipfs/QmYVdQTKmbyz1nU4bR9wCzfqLfmYpnV7zjyPdnno5c8QZF cluster.go:94
16:04:48.503  INFO  consensus: starting Consensus and waiting for a leader... consensus.go:60
16:04:48.504  INFO  consensus: raft cluster is already bootstrapped raft.go:116
16:04:48.505  INFO   ipfshttp: IPFS Proxy: /ip4/127.0.0.1/tcp/9095 -> /ip4/127.0.0.1/tcp/5001 ipfshttp.go:158
16:04:48.505  INFO    restapi: REST API: /ip4/127.0.0.1/tcp/9094 restapi.go:266
16:04:48.505  INFO    cluster: Bootstrapping to /ip4/80.209.230.17/tcp/9096/ipfs/QmVDrbRYZckmBaNYexBw8Kwkhz9fPKGgccN5AARkPSKmrz cluster.go:404
16:05:44.702 ERROR    cluster: context deadline exceeded cluster.go:671
16:05:44.702 ERROR    cluster: context deadline exceeded cluster.go:409
16:05:44.702 ERROR    cluster: Bootstrap unsuccessful cluster.go:132
16:05:44.702  INFO    cluster: shutting down Cluster cluster.go:430
16:05:44.702  INFO  consensus: stopping Consensus component consensus.go:158
16:05:44.702  INFO    monitor: stopping Monitor peer_monitor.go:161
16:05:44.702  INFO    restapi: stopping Cluster API restapi.go:284
16:05:44.702  INFO   ipfshttp: stopping IPFS Proxy ipfshttp.go:532
16:05:44.702  INFO pintracker: stopping MapPinTracker maptracker.go:116
error starting cluster: bootstrap unsuccessful
```

This is using latest master

**Comment by ccampbell** [comment]
One thing… I believe node A is using the code from a couple days ago, and node B is using the latest. I am going to rebuild node A with latest master so they are both using the same code…

**Comment by ccampbell** [comment]
Okay now it fails immediately on node A with: 

```
16:31:04.481 ERROR  consensus: !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! raft.go:162
16:31:04.481 ERROR  consensus: error creating raft: cluster peers do not match raft peers consensus.go:64
16:31:04.481 ERROR    cluster: error creating consensus: cluster peers do not match raft peers cluster.go:184
16:31:04.481  INFO    cluster: shutting down Cluster cluster.go:430
16:31:04.481  INFO    monitor: stopping Monitor peer_monitor.go:161
16:31:04.481  INFO    restapi: stopping Cluster API restapi.go:284
16:31:04.481  INFO    restapi: REST API: /ip4/127.0.0.1/tcp/9094 restapi.go:266
16:31:04.481  INFO   ipfshttp: stopping IPFS Proxy ipfshttp.go:532
16:31:04.481  INFO pintracker: stopping MapPinTracker maptracker.go:116
error starting cluster: cluster peers do not match raft peers
```

I see the raft code has been updated. Is there something special that has to be done in order to start a single node cluster? I would expect it to be able to start up on its own and then other nodes can bootstrap to it. I have peers and bootstrap both set to empty arrays in the config

**Comment by hsanjuan** [comment]
hey @ccampbell please update both nodes to latest master and remove ~/.ipfs-cluster/ipfs-cluster-data. With the latest changes it's better if you start fresh.

**Comment by ccampbell** [comment]
Okay I did all of that. I am still having issues. Node A starts up fine, elects itself as the leader with no other peers.

```
16:50:35.018  INFO    cluster: IPFS Cluster v0.2.1 listening on: cluster.go:92
16:50:35.018  INFO    cluster:         /ip4/127.0.0.1/tcp/9096/ipfs/QmZyixbZdEn4qHj5JBjYP5wnid5ir1yWSoYTsVHFRyC3LM cluster.go:94
16:50:35.018  INFO    cluster:         /ip4/127.0.0.2/tcp/9096/ipfs/QmZyixbZdEn4qHj5JBjYP5wnid5ir1yWSoYTsVHFRyC3LM cluster.go:94
16:50:35.018  INFO    cluster:         /ip4/80.209.230.17/tcp/9096/ipfs/QmZyixbZdEn4qHj5JBjYP5wnid5ir1yWSoYTsVHFRyC3LM cluster.go:94
16:50:35.019  INFO  consensus: starting Consensus and waiting for a leader... consensus.go:60
16:50:35.020  INFO  consensus: raft cluster is already bootstrapped raft.go:116
16:50:35.021  INFO   ipfshttp: IPFS Proxy: /ip4/127.0.0.1/tcp/9095 -> /ip4/127.0.0.1/tcp/5001 ipfshttp.go:158
16:50:35.021  INFO    restapi: REST API: /ip4/127.0.0.1/tcp/9094 restapi.go:266
16:50:37.021  INFO  consensus: Raft Leader elected: QmZyixbZdEn4qHj5JBjYP5wnid5ir1yWSoYTsVHFRyC3LM raft.go:261
16:50:37.021  INFO  consensus: Raft state is catching up raft.go:273
16:50:37.021  INFO  consensus: Consensus state is up to date consensus.go:115
16:50:37.021  INFO    cluster: Cluster Peers (not including ourselves): cluster.go:384
16:50:37.021  INFO    cluster:     - No other peers cluster.go:387
16:50:37.021  INFO    cluster: IPFS Cluster is ready cluster.go:394
```

But when I try to bootstrap Node B to Node A I get this on Node B:

```
16:52:23.922 ERROR    cluster: dial attempt failed: misdial to <peer.ID VDrbRY> through /ip4/80.209.230.17/tcp/9096 (got <peer.ID ZyixbZ>): read tcp 80.209.230.16:9096->80.209.230.17:9096: use of closed network connection cluster.go:671
16:52:23.922 ERROR    cluster: dial attempt failed: misdial to <peer.ID VDrbRY> through /ip4/80.209.230.17/tcp/9096 (got <peer.ID ZyixbZ>): read tcp 80.209.230.16:9096->80.209.230.17:9096: use of closed network connection cluster.go:409
16:52:23.922 ERROR    cluster: Bootstrap unsuccessful cluster.go:132
16:52:23.922  INFO    cluster: shutting down Cluster cluster.go:430
16:52:23.922  INFO  consensus: stopping Consensus component consensus.go:158
16:52:23.922  INFO    monitor: stopping Monitor peer_monitor.go:161
16:52:23.922  INFO    restapi: stopping Cluster API restapi.go:284
16:52:23.922  INFO   ipfshttp: stopping IPFS Proxy ipfshttp.go:532
16:52:23.922  INFO pintracker: stopping MapPinTracker maptracker.go:116
```

I am happy to give you the cluster secret if you want to try to join

**Comment by hsanjuan** [comment]
@ccampbell can you check you are bootstrapping to the right multiaddress ? From that message, it seems the peer ID in it is wrong

**Comment by ccampbell** [comment]
Okay it all seems to be working now! I will do some more testing and let you know if I discover anything. 🤞

I tried AB cluster and killed each node one at a time and brought them back up and they successfully rejoined

**Comment by ccampbell** [comment]
Okay now I am seeing a different issue. This one may be a problem between the chair and keyboard, but I am seeing that after adding a file and pinning it to the cluster, it is hanging and eventually returning a 504 gateway timeout when requesting it from any of the nodes in the cluster that are not the node I added.

In this scenario I have my original cluster A, B, C (One of them is the leader). I joined the cluster with a new node D then ran the following from node D:

```
ipfs add file.txt
ipfs-cluster-ctl pin add [hash of file]
``` 

### ipfs-cluster logs

- Node A shows nothing
- Node B shows:

    ```
    pin committed to global state: [hash]
    ```
- Node C shows nothing
- Node D shows:

    ```
    IPFS cluster pinning [hash] on [<peer.ID WdLkGA> <peer.ID ZyixbZ>]: cluster.go:854
    ```

I was under the impression that you could request pinned content from any node in the cluster and they should be able to return it and cache it locally? Previously when I tested this, it seemed to work fine. Am I doing something wrong?

Thanks

**Comment by hsanjuan** [comment]
Hmm, it seems that steps on the ipfs side of things. Do your ipfs daemons have regular connectivity? Are they connected directly to the deamons of other cluster peers (`ipfs swarm peers` should show the multiaddresses of the others...) ? cluster is not involved in fetching content from ipfs, it just tells ipfs to pin. If A,B,C hang, it looks like those nodes cannot find providers for the file you just added. Sometimes, just restarting the ipfs daemon does the trick.

**Comment by ccampbell** [comment]
Hmm they definitely have regular connectivity. I may have messed up the config. I am trying some things, but now when I run `ipfs swarm peers` the cluster service immediately crashes after logging:

```
13:11:18.915 ERROR    cluster: consensus start timed out cluster.go:375
13:11:18.915  INFO    cluster: shutting down Cluster cluster.go:430
13:11:18.915  INFO  consensus: stopping Consensus component consensus.go:158
13:11:18.915  INFO    monitor: stopping Monitor peer_monitor.go:161
13:11:18.915  INFO    restapi: stopping Cluster API restapi.go:284
13:11:18.915  INFO   ipfshttp: stopping IPFS Proxy ipfshttp.go:532
13:11:18.915  INFO pintracker: stopping MapPinTracker maptracker.go:116
13:11:18.915 ERROR    cluster: connection reset cluster.go:605
```

This is a side note, but is there a way to make ipfs nodes only discoverable to nodes within the same cluster but not to the public internet?

**Comment by ccampbell** [comment]
`ipfs swarm peers` is no longer crashing after restarting everything, but it is still returning nothing. All the peers are there when I run `ipfs-cluster-ctl peers ls` from each node

**Comment by hsanjuan** [comment]
hey @ccampbell , `ipfs swarm peers` and cluster have no relation. The log messages you posted seem from a cluster peer which failed to start properly (maybe because other peers weren't ready).

if you add something to ipfs and other ipfs nodes cannot see it, you can debug it with cluster turned off as it plays no part.

Did you remove the bootstrap nodes from the ipfs configuration? if `ipfs swarm peers` comes back empty, it means that ipfs does not see any other nodes, therefore DHT lookups will fail unless the content is pinned locally.

> This is a side note, but is there a way to make ipfs nodes only discoverable to nodes within the same cluster but not to the public internet?

See https://github.com/ipfs/go-ipfs/blob/master/docs/experimental-features.md#private-networks

**Comment by ccampbell** [comment]
Thanks. I did in fact remove the bootstrap nodes from the ipfs configuration. For some reason, I expected that by joining a cluster using bootstrap, that would also communicate to the ipfs daemon to connect to the other nodes from the cluster. Would that behavior be possible? 

I suppose it would be possible to set up something similar using the private network key you just linked to, but it would be cool if this was supported out of the box. 

Anyway I will try some more things, and let you know if I run into anything else. Sorry for all the back and forth!

**Comment by hsanjuan** [comment]
> For some reason, I expected that by joining a cluster using bootstrap, that would also communicate to the ipfs daemon to connect to the other nodes from the cluster. Would that behavior be possible? 

In fact, cluster does tell ipfs to `ipfs swarm connect` to other ipfs nodes in the cluster, both when starting a fixed cluster and when adding a new peer. I'll have to check this again, since it would seem it's not quite working for you. Are you starting cluster and ipfs at the same time when you bootstrap in a way that the ipfs daemon would not be 100% ready while bootstrapping cluster?

**Comment by ccampbell** [comment]
The way I have been testing is

- Start ipfs daemon on node A with bootstrap empty
- Start ipfs cluster on node A with cluster secret but no bootstrap
- Start ipfs daemon on node B with bootstrap empty
- Start ipfs cluster on node B bootstrapping to node A

So to answer your question, I am always making sure the daemon is running before running the cluster service

**Comment by ccampbell** [comment]
I am having trouble trying to set up a private network too 😕. I have copied the same swarm.key to nodes A, B, and C. I then removed the bootstrap and started ipfs daemon node A (without bootstrap) and then on node B ran

```
ipfs bootstrap rm --all
ipfs bootstrap add <multiaddr of node A>
ipfs daemon
```

It starts up the daemon, but ipfs swarm peers still shows nothing. The swarm key fingerprint matches during the startup logs.

**Comment by hsanjuan** [comment]
@ccampbell I have tried to do this (set up ipfs with swarm key) and it works for me :/

I am using the latest stable 0.4.11. I have 4 ipfs nodes. I have followed the instructions, and then added the multiaddr of node A to everyone (including itself, for convenience). Content added in one is retrievable from the other.

Can you check the logs for any errors? Can you verify connectivity between those ipfs nodes? Can you double check that you used the right multiaddress?

**Comment by ccampbell** [comment]
Okay. I have done some more testing, and I have figured out my issue. It turned out to be something dumb 😞 . I did not have port 4001 open on the firewall for the nodes. 9096 was open for the cluster but not 4001 for the daemon. 

So there are two situations where I think the output could be improved where I would have been able to figure out the problem right away:

1. If you have a private swarm.key for an ipfs node and you try to bootstrap to another ipfs node it should give an error if that node is unreachable. Perhaps something like:

```
Error bootstrapping to peer with multiaddr: <multiaddr>, please make sure ipfs daemon is running and port 4001 is open on your firewall
```

2. If your ipfs daemon has no bootstrap nodes and you create a cluster and run ipfs-cluster-service with --bootstrap set to another node, it should give an error if the local daemon is not able to connect to the daemon on the other cluster node (it currently errors if the other cluster peer can’t be reached during bootstrap, but not for daemon peers). 

Could be the same error message or maybe something like this to point out that the problem is in the daemon connection vs. the cluster

```
Error bootstrapping ipfs daemon to peer with multiaddr…
```

---

I could be wrong, but I am pretty sure that both situations failed silently which is what led to me running into most of these issues. The daemon started up just fine and said it was ready, but nothing was able to talk to it cause of the port being closed.

Thanks for being so responsive. I **think** things are looking pretty good right now.

**Comment by hsanjuan** [comment]
Hey! Glad things are looking better.

Point 1 is a pure go-ipfs thing so you should probably report it there. But ipfs makes no assumptions about other nodes' availability. Perhaps this should be different if no bootstrap node works or swarm+empty bootstrap.

For point2, if you run cluster in debug mode, you will see errors about `POST swarm/connect` from the `ipfshttp` component. This is when cluster tries to link ipfs daemons. The problem is that ipfs daemons advertise multiple addresses (external, local network, localhost) so we try to connect on them all, and some of them always throw an error. It was very confusing to spit errors on every boot so I left it as debug information.

That said @ZenGround0 was working on some tooling for identifying connection problems (#17 ) and we also have proposed things like #165 .

**Comment by ccampbell** [comment]
Hey there. It has been a while since I have commented on this ticket. Things had been running okay, but lately the cluster that had been running has been having a number of issues, and I am not even sure where to begin. There have been issues with the health of some of the nodes which prevents new nodes from joining. 

```
11:18:04.047 ERROR    cluster: error broadcasting new peer's address: all cluster members need to be healthy for this operation to succeed. Try removing any unhealthy peers. Check the logs for more information about the error. cluster.go:775
```

The bad nodes for whatever reason are not removed from the peers list from all nodes even though `leave_on_shutdown` is set to true. Sometimes manually removing them from the node I am bootstrapping to fixes it, but sometimes it does not. I think at one point the bad node happened to be the leader, and I ran `ipfs-cluster-ctl peers rm [id]` not realizing that, and maybe that magnified the issue.

---

There were also a few nodes connected to the cluster that one day, out of the blue, disconnected, and have not been able to join since. I have tested with a brand new install on a new cluster, and they are not able to connect to that one either. Could it be some system level thing that needs to be reset? I tried rebooting the problematic machine(s) and it did not seem to help. The error here is:

```
11:26:43.808  INFO  consensus: starting Consensus and waiting for a leader... consensus.go:61
11:26:43.810  INFO  consensus: raft cluster is already bootstrapped raft.go:119
11:26:43.813  INFO    cluster: Bootstrapping to /ip4/52.71.207.22/tcp/9096/ipfs/QmbNnPnCzVF2tfamBBmn6Lqy59cWKNwEaoYT1NwoTS2iFS cluster.go:463
11:26:43.814  INFO   ipfshttp: IPFS Proxy: /ip4/127.0.0.1/tcp/9095 -> /ip4/127.0.0.1/tcp/5001 ipfshttp.go:159
11:26:43.815  INFO    restapi: REST API: /ip4/127.0.0.1/tcp/9094 restapi.go:272
11:26:44.963 ERROR    cluster: protocol not supported cluster.go:775
11:26:44.963 ERROR    cluster: protocol not supported cluster.go:468
11:26:44.963 ERROR    cluster: Bootstrap unsuccessful cluster.go:133
11:26:44.963  INFO    cluster: shutting down Cluster cluster.go:490
11:26:44.963  INFO  consensus: stopping Consensus component consensus.go:159
11:26:44.963 ERROR       raft: Failed to take snapshot: nothing new to snapshot logging.go:86
11:26:44.963  INFO    monitor: stopping Monitor peer_monitor.go:161
11:26:44.963  INFO    restapi: stopping Cluster API restapi.go:290
11:26:44.963  INFO   ipfshttp: stopping IPFS Proxy ipfshttp.go:534
11:26:44.964  INFO pintracker: stopping MapPinTracker maptracker.go:105
error starting cluster: bootstrap unsuccessful
```

It is failing during the bootstrap for some reason, but I haven’t spent a lot of time debugging it.

This is had been happening using an older version of ipfs-cluster code from December, so I started a new cluster using a build of master from last week, and it is still happening.

---

One other thing I am wondering about is related to the IPFS swarm peers. The new cluster I made has 5 nodes currently in it, but the `ipfs swarm peers` command only shows 1 or 2 or 3 on each machine. I would expect it to show all.

Also for two nodes that I launched from AWS EC2, I noticed that the public IP v4 address is not listed in the `ipfs-cluster-ctl peers ls`, instead it lists the internal IP from EC2. For example:

```
QmQYvC588snLp5NE7wpBvNh8jskim8ZeSzHacq2hvp5A3X | ip-172-31-42-34 | Sees 3 other peers
  > Addresses:
    - /ip4/127.0.0.1/tcp/9096/ipfs/QmQYvC588snLp5NE7wpBvNh8jskim8ZeSzHacq2hvp5A3X
    - /ip4/172.31.42.34/tcp/9096/ipfs/QmQYvC588snLp5NE7wpBvNh8jskim8ZeSzHacq2hvp5A3X
  > IPFS: QmXjAqwVeZm8JcZJtETy7fnu8RdJjqj8B2hzA6nL5mMLXu
    - /ip4/127.0.0.1/tcp/4001/ipfs/QmXjAqwVeZm8JcZJtETy7fnu8RdJjqj8B2hzA6nL5mMLXu
    - /ip4/172.31.42.34/tcp/4001/ipfs/QmXjAqwVeZm8JcZJtETy7fnu8RdJjqj8B2hzA6nL5mMLXu
    - /ip6/::1/tcp/4001/ipfs/QmXjAqwVeZm8JcZJtETy7fnu8RdJjqj8B2hzA6nL5mMLXu
```

Is there a way to get it to display the public IP v4 and could that be related to why these nodes are not showing up in the swarm peers for other nodes in the cluster?

Sorry, that is a bunch of stuff. I don’t really know if there is a better way you would prefer to communicate.

Thanks so much

**Comment by hsanjuan** [comment]
Hey!

>  The bad nodes for whatever reason are not removed from the peers list from all nodes even though leave_on_shutdown is set to true.

Sure, if the node is gone you need to `ipfs-cluster-ctl peers rm <peerID>` manually. `leave_on_shutdown` only removes on an orderly shutdown of the peer.

> I think at one point the bad node happened to be the leader

If a node becomes unresponsive, a new leader will be chosen. If the leader is remove, the remaining peers will choose a new one.

> Is there a reason that new nodes are not able to join if one of the existing peers is in a bad state?

Yes, we need to inform peers of the location (addresses) of new the Peer. If a peer is down, you add a new peer, and then that peer comes up, it doesn't know where the new peer is. Can we work around this? Yes, sharing peers addresses probably, opportunistically. A peer discovery mechanism would help too (consul #158  or mdns), but we don't have that (and doesn't work for all setups).

If a peer doesn't know how to contact another peer you are left with a very dysfunctional cluster which is even more confusing than this early clear error.

So short story long, better to fail early and clearly now, until this part can be improved.

---

> 11:26:44.963 ERROR    cluster: protocol not supported cluster.go:775

This is for sure telling you that you are running cluster on different versions and that two nodes with a different version fail to speak among themselves.

---

> One other thing I am wondering about is related to the IPFS swarm peers. 

`connect_swarms_delay` how long after a peer comes up, it's ipfs daemon will be instructed to connect to the ipfs daemons of all other peers, so this sure ensure that peers appear listed `swarm peers`.

If the other peers' daemons don't show the right address, however, this won't work. I don't know about your setup, but it would seem that ipfs does not realize it has an external ip (is it a private-network?). Normally this is second-guessed when ipfs contacts the bootstrap servers (assuming this happens over the external external ip).  Otherwise, ipfs will only list what it can see locally (I guess `172.31.42.34`). So it depends on your ipfs config and aws setup...

**Comment by ccampbell** [comment]
Thanks for the fast reply. I will look into the AWS IP address thing. Not too familiar with how that works.

I just checked the protocol not supported though and I am certain they are running the same code. I did a shasum on the `ipfs-cluster-service` binary on the peer im trying to connect from and the peer already running that I am bootstrapping to and they both match.

Are there any files on disk maybe that need to be cleaned up with a new cluster version that exist outside of `~/.ipfs-cluster` or `~/.ipfs`? 

I tried completely removing those directories, put the new binary, and then tried to start it bootstrapping to a running node, and it gives the error

**Comment by hsanjuan** [comment]
> I tried completely removing those directories, put the new binary, and then tried to start it bootstrapping to a running node, and it gives the error

And other nodes?

**Comment by hsanjuan** [comment]
And which version are you running?

**Comment by ccampbell** [comment]
Yes, I just verified all nodes are running the same one. I don’t know if this shows the current git revision, but:

```
0.3.4-4e1f59062d64b8816b2c883124205e362e9d6e6c
```

This issue was happening with older code too. These specific problematic nodes were connected to the cluster and everything was working fine, then one day the processes crashed, and ever since they have not been able to connect again. I thought I would try creating a brand new cluster with the updated code, removed the directories, and tried again, but same issue. 

Could it have to do with other software running on the box/specific ports that are causing problems?

**Comment by ccampbell** [comment]
Yes, it is. I *just* set up these nodes recently. A brand new node is able to connect fine, and other existing nodes are able to connect fine too. There are just 3 or 4 problematic ones that can’t seem to connect to any cluster no matter what

**Comment by hsanjuan** [comment]
@ccampbell I reproduce exactly the same error messages when bootstrapping a peer against another peer running a different cluster version.. this is the only thing I can think about that can produce this error.

Running with the commit you posted:

```
18:24:18.585  INFO    cluster: IPFS Cluster v0.3.4-4e1f5906 listening on: cluster.go:90
18:24:18.585  INFO    cluster:         /ip4/127.0.0.1/tcp/9096/ipfs/Qmb9qpF5JkxkJWfmoXydwjATbhBwfp2uF8yVC8BbhyXpPd cluster.go:95
18:24:18.585  INFO    cluster:         /ip4/192.168.123.103/tcp/9096/ipfs/Qmb9qpF5JkxkJWfmoXydwjATbhBwfp2uF8yVC8BbhyXpPd cluster.go:95
18:24:18.589  INFO    cluster: Bootstrapping to /ip4/127.0.0.1/tcp/10096/ipfs/Qma7KpVhrTgkdUxBanyVSffjydx9dceDXDz1K8GzspB7me cluster.go:469
18:24:18.589  INFO   ipfshttp: IPFS Proxy: /ip4/127.0.0.1/tcp/9095 -> /ip4/127.0.0.1/tcp/5001 ipfshttp.go:171
18:24:18.589  INFO    restapi: REST API: /ip4/127.0.0.1/tcp/9094 restapi.go:285
18:24:18.606 ERROR    cluster: protocol not supported cluster.go:781
18:24:18.606 ERROR    cluster: protocol not supported cluster.go:474
18:24:18.606 ERROR    cluster: Bootstrap unsuccessful cluster.go:137
18:24:18.606  INFO    cluster: shutting down Cluster cluster.go:496
18:24:18.606  INFO  consensus: stopping Consensus component consensus.go:159
18:24:20.408 ERROR       raft: Failed to take snapshot: nothing new to snapshot logging.go:86
18:24:20.408  INFO    monitor: stopping Monitor peer_monitor.go:161
18:24:20.408  INFO    restapi: stopping Cluster API restapi.go:303
18:24:20.408  INFO   ipfshttp: stopping IPFS Proxy ipfshttp.go:548
18:24:20.408  INFO pintracker: stopping MapPinTracker maptracker.go:105
error starting cluster: bootstrap unsuccessful
```

The only difference are the line for the errors. If lines are different,  you must be running code other than `4e1f59062d64b8816b2c883124205e362e9d6e6c`?

**Comment by ccampbell** [comment]
Aha! The issue is that I have two binaries in different locations in ~/bin and in ~/go/bin and ~/go/bin is first in my $PATH ughhh. 😥 Thank you.

**Comment by hsanjuan** [comment]
Glad you found it!

**Comment by ccampbell** [comment]
Hello again. Sorry to bother you, but I have another head-scratcher now. I am trying to get a node with a clean install to join an existing cluster and having trouble.

It seems to connect just find (bootstrapping to another node) and find all the peers, but then it mysteriously shuts down without an error:

```
20:15:00.960  INFO    cluster: Bootstrapping to /ip4/45.77.136.112/tcp/9096/ipfs/QmZXjPy712cuepoiFEFFPLKCx13M22hULNPG7QuvnySTUr cluster.go:135
20:15:00.960  INFO   ipfshttp: IPFS Proxy: /ip4/127.0.0.1/tcp/9095 -> /ip4/127.0.0.1/tcp/5001 asm_amd64.s:2361
20:15:00.961  INFO    restapi: REST API: /ip4/127.0.0.1/tcp/9094 asm_amd64.s:2361
20:15:01.960  INFO  consensus: Current Raft Leader: QmNxZPCpPsTYjbFkJA24MnGdFijSgXpQjpUyATsPK49PEt consensus.go:98
20:15:06.084  INFO  consensus: Current Raft Leader: QmNxZPCpPsTYjbFkJA24MnGdFijSgXpQjpUyATsPK49PEt consensus.go:98
20:15:06.104  INFO    cluster: QmdJafhiDDdaKNfdUBHdZa7QEJbZDPqfo2FEsmUnyVx8NF: joined QmZXjPy712cuepoiFEFFPLKCx13M22hULNPG7QuvnySTUr's cluster cluster.go:470
20:15:06.105  INFO    cluster: Cluster Peers (without including ourselves): cluster.go:142
20:15:06.105  INFO    cluster:     - QmNxZPCpPsTYjbFkJA24MnGdFijSgXpQjpUyATsPK49PEt cluster.go:142
20:15:06.105  INFO    cluster:     - QmPYVEGvs4ju78VdksKTuUcPHg9impCxWiGbqo7hLMm7GV cluster.go:142
20:15:06.105  INFO    cluster:     - QmUqb7tmHvjp3bfNc7iAsX7S8K7bc244cpCuEvUZyZGURS cluster.go:142
20:15:06.105  INFO    cluster:     - QmWTay6Rj5dzmokSSvAddBcqBiZe58AhDkw4roEsGq46cE cluster.go:142
20:15:06.105  INFO    cluster:     - QmXpz22zLY3WBkrv6g7Xj5RhoJhKrxkhYjBcQQKp3bcGDq cluster.go:142
20:15:06.105  INFO    cluster:     - QmZXjPy712cuepoiFEFFPLKCx13M22hULNPG7QuvnySTUr cluster.go:142
20:15:06.105  INFO    cluster:     - QmaZ87gi19TitqMXydgzks3t4CU5pGr5gmZnPXegSf8BSu cluster.go:142
20:15:06.105  INFO    cluster: ** IPFS Cluster is READY ** cluster.go:142
20:15:06.950  INFO     config: Saving configuration config.go:130
20:15:11.108  INFO    cluster: QmdJafhiDDdaKNfdUBHdZa7QEJbZDPqfo2FEsmUnyVx8NF: removed from raft. Initiating shutdown asm_amd64.s:2361
20:15:11.116  INFO    cluster: shutting down Cluster asm_amd64.s:2361
20:15:11.116  INFO  consensus: stopping Consensus component cluster.go:519
20:15:11.131  INFO  consensus: consensus data cleaned cluster.go:534
20:15:11.131  INFO    monitor: stopping Monitor cluster.go:540
20:15:11.131  INFO    restapi: stopping Cluster API cluster.go:545
20:15:11.131  INFO   ipfshttp: stopping IPFS Proxy cluster.go:549
20:15:11.131  INFO pintracker: stopping MapPinTracker cluster.go:554
20:15:11.132 ERROR libp2p-raf: Failed to decode incoming command: connection reset log.go:172
20:15:11.949  INFO     config: Saving configuration config.go:130
```

This seems to happen every time even after a clean installation. I traced the shutdown to the `Cluster.watchPeers()` method. It seems like even though this peer was bootstrapped to the cluster the other peers are not able to see it? 

Do you have any recommendation of what to do to debug this?

Thanks so much

**Comment by hsanjuan** [comment]
@ccampbell:

Can you post the `--loglevel debug` logs for this peer?

Can you post the logs of the peer this node is bootstrapping to?

Can you post your peer configuration file (without the private key and secret)?

**Comment by ccampbell** [comment]
Also not sure if it is related, but I just stopped a peer and started it again and it seemed to connect okay (a different peer), but there are a lot of these in the logs:

```
17:56:09.850 ERROR pintracker: pin queue is full rpc_api.go:207
17:56:09.850 ERROR  p2p-gorpc: pin queue is full asm_amd64.s:2361
```

It showed up hundreds of times

**Comment by hsanjuan** [comment]
> pin queue is full

Hm, I think this is not critical, but I'll consider it a bug (#377). You can try to increase `pintracker.max_pin_queue_size` as a workaround. I understand you have a pinset with more than 4096 elements?

Regarding your `peer removed` problem: I am not sure. The peer seems to correctly bootstrap and correctly syncs the raft state, but from your logs, I think the raft state may not be fully synced after 5 seconds (#378).

If I am correct on my assumptions, there might be a work around. In cluster `0.3.5` we introduced a `cluster.peer_watch_interval` option, which defaults to 5 seconds. If you upgrade your cluster, and set this option to, say, `300s` (in the new peer's config), then your new peer should hopefully not auto-shutdown (or do it after 300 seconds, in which case I have to look for another cause). Unfortunately you cannot change this on the versions you're using (`0.3.4`).

Also, @ccampbell, thank you very much for reporting all this. This feedback is very valuable. Could you tell me:

* How many peers you have in your cluster?
* How big is the pinset?
* How often do you pin/unpin things?

I'd like to replicate your setup (or at least the raft-log size).

**Comment by ccampbell** [comment]
Hi, thanks for your reply. I am working on this. Updating to 0.3.5 seems like it has a new state version so I have to update each node one at a time. 

I didn’t realize that there is a limit on how many things can be pinned in a cluster. Now I am pretty confused cause I went to pin something new and I got this:

```
$ ipfs-cluster-ctl pin add  QmUxyrovj8U7Fv78pN1aBsoCQHoT9TQdJCN17dmf1sBbWM
QmUxyrovj8U7Fv78pN1aBsoCQHoT9TQdJCN17dmf1sBbWM :
    > Peer QmNxZPCpPsTYjbFkJA24MnGdFijSgXpQjpUyATsPK49PEt : ERROR | pin queue is full
    > Peer QmPYVEGvs4ju78VdksKTuUcPHg9impCxWiGbqo7hLMm7GV : REMOTE | 2018-04-18T19:48:59Z
    > Peer QmUqb7tmHvjp3bfNc7iAsX7S8K7bc244cpCuEvUZyZGURS : ERROR | pin queue is full
    > Peer QmWTay6Rj5dzmokSSvAddBcqBiZe58AhDkw4roEsGq46cE : ERROR | pin queue is full
    > Peer QmXpz22zLY3WBkrv6g7Xj5RhoJhKrxkhYjBcQQKp3bcGDq : ERROR | pin queue is full
    > Peer QmZXjPy712cuepoiFEFFPLKCx13M22hULNPG7QuvnySTUr : REMOTE | 2018-04-18T19:48:59Z
    > Peer QmaZ87gi19TitqMXydgzks3t4CU5pGr5gmZnPXegSf8BSu : ERROR | pin queue is full
```

As for your questions, you can see there are only 7 peers at the moment. As for the pinning, I recently set up a process to check the cluster health using files added to the cluster. We were experiencing issues where certain ipfs-cluster daemon’s would crash (not sure why) and they would no longer be a part of the cluster, but the IPFS daemon would still be serving existing content. So it meant new pins were not able to be delivered by certain nodes. 

So what I did was I set up a process that creates a new “health check” file and pins it to the cluster every 5 minutes, then it unpins the old one. After that it monitors the peers to make sure that they can deliver the new file. It seemed to be working fine, but I guess at some point the pin queue got too large and it stopped working? 

Does the pin queue keep track of all unpin and pin requests the cluster has ever received? If so, is there any way to achieve something like what I described?

**Comment by hsanjuan** [comment]
The problem filling the 4092 pin queue completely is probably that some pin is stuck and nothing is getting is through, so it eventually fills up. I wouldn't be surprised if these are health checks files not getting through, as otherwise your pin set is really small.

We are fixing a bunch of things in the pintracker (#308) to address some of these issues, but in the meantime:

> So what I did was I set up a process that creates a new “health check” file and pins it to the cluster every 5 minutes, then it unpins the old one. After that it monitors the peers to make sure that they can deliver the new file

ipfs will always be able to deliver the file if it's somewhere, via DHT query. The question is if the cluster peer has pinned the file correctly.


> Updating to 0.3.5 seems like it has a new state version so I have to update each node one at a time.

You just need to install the new version and run your peers with `ipfs-cluster-service daemon --upgrade`. Remember peers won't talk to each other when running on different versions.

**Comment by ccampbell** [comment]
> The problem filling the 4092 pin queue completely is probably that some pin is stuck and nothing is get

Is there a workaround for this now or some way to clear the current pin queue manually?

> Remember peers won't talk to each other when running on different versions.

I’m aware. I read that in the release notes. Are there any other steps that need to be done for a smooth upgrade or will it be okay to upgrade one node at a time and stop and bootstrap each node to one running the 0.3.5 version? Do I need to generate a new cluster secret or ipfs swarm.key first?

One other very minor note. When you try to run the new version to bootstrap to an old version you get the “protocol not supported” error which is great. It has now been updated to include a list of reasons for why it might be happening, but none of the reasons in the list is that “the peer you are trying to start is running a different version of IPFS-cluster”. I feel like may end up being the most common reason. Each time I’ve run into it so far that was why. So might be worth adding it to the error.

Thanks again.

**Comment by ccampbell** [comment]
So on the first node I am trying to upgrade I got this error:

```
error upgrading state: no snapshot found
```

It seems to me that snapshots should be taken automatically. I did notice it looks like I had one in one of the `~/.ipfs-cluster/ipfs-cluster-data.old.0` or `ipfs-cluster-data.old.1` directories. 

Now I started up the new cluster version on a node without bootstrapping to the cluster running the previous version so I think it completely removed the state. Now when I try to bootstrap to the cluster running 0.3.4 from the peer I am getting this error:

```
error creating REST API component:  is not a 'thin waist' address
```

I think I may try removing the ~/.ipfs-cluster directory completely and starting over completely to try to join the current cluster (0.3.4) to see if a snapshot is generated so I can upgrade. Does that seem right to you?

**Comment by ccampbell** [comment]
Also maybe it was my own fault. Are you supposed to run the upgrade command while the ipfs-cluster-service daemon is running? I stopped it with kill -HUP before running the upgrade command

**Comment by ccampbell** [comment]
Okay it seems like as soon as the cluster is not running the `ipfs-cluster-data` directory is renamed `ipfs-cluster-data.old.0` and the upgrade command is not able to find the snapshots.

**Comment by hsanjuan** [comment]
Upgrades are supposed to be run with daemon offline. If there is no snapshot, then there is nothing to ugprade and you can just start the new version of cluster. 

Now, if there is no state, that can be because you removed the peer from the cluster (in that case state is cleaned up for you). Since you are bootstrapping those peers then they should not have state so I think everything's good (sorry, I completely forgot the state upgrade part isn't necessary on peers without state and that that was your case).

Check your config, about the `thin waist` address in the `restapi` section. There must be something not set correctly (some of the listen multiaddresses).

**Comment by ccampbell** [comment]
So just to summarize. When I kill -HUP on this node and the other nodes remove it from the cluster (leave_on_shutdown is set to true), the `ipfs-cluster-data` directory (which has a snapshot) gets renamed to `ipfs-cluster-data.old.0` and an upgrade is not needed to start up using the new version?

There still is a snapshot in there it is just no longer in `ipfs-cluster-data/snapshots`.

**Comment by hsanjuan** [comment]
> So just to summarize. When I kill -HUP on this node and the other nodes remove it from the cluster (leave_on_shutdown is set to true), the ipfs-cluster-data directory (which has a snapshot) gets renamed to ipfs-cluster-data.old.0 and an upgrade is not needed to start up using the new version?

Yes, correct!

**Comment by ccampbell** [comment]
Okay great. I will give it a go. My plan is something like this:

- current cluster: peer A - peer B - peer C (all running 0.3.4)
- remove peer C from cluster -> upgrade to 0.3.5 -> start peer C on its own
- remove peer B from cluster -> upgrade to 0.3.5 -> bootstrap to peer C
- stop peer A -> upgrade to 0.3.5 -> bootstrap to peer C

Does that sound right? And all the pinned content should remain?

**Comment by hsanjuan** [comment]
NO

Do this:

* Remove peer C -> upgrade to 0.3.5
* Remove peer B -> upgrade to 0.3.5
* Stop peer A -> upgrade to 0.3.5 -> `state upgrade` -> start
* Bootstrap peer C to peer A
* Bootstrap peer B to peer A

If you want to keep the state, you should not bootstrap to a peer which has cleaned it after being removed.

**Comment by ccampbell** [comment]
Ahh so the final node needs the state upgrade, but that means that for some period of time the entire cluster has to be down? Is there any way to maintain 100% uptime during an upgrade?

I suppose at the least, I can keep IPFS daemon running which means delivery can continue to work

**Comment by ccampbell** [comment]
Ideally there would be a way to remove one node, do a state upgrade on that and then bootstrap all the other nodes one at a time to the new one. If the network has 100 nodes, having to remove all 99 just to get to the final one to update the state and then bootstrap again seems a bit cumbersome unless I am missing something

**Comment by ccampbell** [comment]
Hmm. Well I believe the issue I was seeing with that was that the peer list does not get updated. That means as soon as a node is removed with `remove_on_shutdown: false`, any other node that tries to bootstrap into the cluster will fail since the cluster itself will be in an “unhealthy” state. So I would have to manually go in and remove the nodes that are timing out with `ipfs-cluster-ctl peers rm [hash]` on the bootstrap node. 

Why would it require more leader elections to remove one node and then slowly bootstrap the old nodes to the new one one at a time? It seems like the suggestion now would have to swap out leaders each time a node is removed until the cluster is down to the final node at which point the state can be upgraded and the other nodes can bootstrap back in.

Setting `remove_on_shutdown: false` sounds more appealing in terms of the upgrade process, but is there a way to do that while also allowing nodes to join and making sure the cluster state is up to date and healthy?

**Comment by hsanjuan** [comment]
> Setting remove_on_shutdown: false sounds more appealing in terms of the upgrade process, but is there a way to do that while also allowing nodes to join and making sure the cluster state is up to date and healthy?

Cluster will be healthy as long as a majority of peers is still up. After all, `remove_on_shutdown` only has effect on clean shutdowns of the peer. So in the case of peers dying suddenly, it does not help in keeping the cluster healthy.

I do agree things are not optimal here. Next round of iteration over raft should improve this (3rd point here: https://github.com/ipfs/ipfs-cluster/issues/384). The underlying problem is that Raft imposes many constraints on how things should work in terms of peerset modifications in order to keep the consensus working.

**Comment by ccampbell** [comment]
Mmm. Well glad it is on the radar. Thanks for all the fast responses. I will let you know if I run into any other hiccups.

**Comment by NatoBoram** [comment]
Adding a `ipfs-cluster-service shutdown` command would allow `remove_on_shutdown` to be performed if the cluster is running in the background

**Comment by hsanjuan** [comment]
@NatoBoram this would be roughly the equivalent of `ipfs-cluster-ctl peers rm <peer-id>`. Peers shutdown when removed. And you can just down/remove any peer, not just self.

**Comment by NatoBoram** [comment]
If I understand correctly, the correct way to cleanly shutdown the service when it's running as a service is to have it remove itself from the cluster?

**Comment by hsanjuan** [comment]
You can shut-down any peer at any time. But depending on why you shut it down and on the state of your cluster, you may want to remove it from the consensus peerset so that the rest of peers don't assume that the peer is going through an abnormal situation and keep trying to contact it. `remove_on_shutdown` does that automatically. Also, removing a peer with `peers rm` will cause it to shutdown afterwards.

But in all cases, ctrl-c'ing or stopping the service is a clean shutdown from the local point of view.

**Comment by hsanjuan** [comment]
Since we merged #685 , I'm closing this as fixed. With `--consensus crdt` peers can now join and leave without impacting other peers.

**Comment by NatoBoram** [comment]
Where can we use `--consensus crdt`?

**Comment by kishansagathiya** [comment]
`ipfs-cluster-service daemon --consensus crdt`

**Comment by NatoBoram** [comment]
```bash
$ ipfs-cluster-service daemon --consensus crdt

Incorrect Usage: flag provided but not defined: -consensus

NAME:
   ipfs-cluster-service daemon - run the IPFS Cluster peer (default)

USAGE:
   ipfs-cluster-service daemon [command options] [arguments...]

OPTIONS:
   --upgrade, -u                run necessary state migrations before starting cluster service
   --bootstrap value, -j value  join a cluster providing an existing peers multiaddress(es)
   --alloc value, -a value      allocation strategy to use [disk-freespace,disk-reposize,numpin]. (default: "disk-freespace")
   --stats                      enable stats collection
   --tracing                    enable tracing collection

```

Version `0.10.1`, built from `master`. What branch should I use?

**Comment by hsanjuan** [comment]
The official way is documented here: https://cluster.ipfs.io/download/#installing-from-source

**Comment by NatoBoram** [comment]
I couldn't build it using the old method, and that's where I got the idea of building it like that. I followed those instructions but ended up with the wrong version, `0.10.1`, so I had to search again for how to build it.

**Comment by hsanjuan** [comment]
actually you're right, it doesn't set the commit information if you build directly, but you should get a master build if you're on master. `version` will say `0.10.1` because that's the last thing we released.

**Comment by NatoBoram** [comment]
But when I ran `ipfs-cluster-service daemon --consensus crdt`, I got https://github.com/ipfs/ipfs-cluster/issues/160#issuecomment-487340241, so I'm pretty sure I didn't get the right version. I'm not sure how that happened though, since I was building in the `master` branch.

**Comment by eiTanLaVi** [comment]
Is there a simple way to check which peer is the current leader?

**Comment by lanzafame** [comment]
@eiTanLaVi no. But for what purpose?

**Comment by eiTanLaVi** [comment]
to checksum leader election process has settled, part of cluster health checks etc..

**Comment by eiTanLaVi** [comment]
Q: if we use the --upgrade flag, on v0.10.1, will the --consensus crdt flag take affect when restarting the daemon (systemd) ? what is the safe way to upgrade to this flag on a running cluster?

**Comment by hsanjuan** [comment]
Answered in #685 . Please don't double post

**Comment by hsanjuan** [comment]
> to checksum leader election process has settled, part of cluster health checks etc..

it sometimes will appear on the logs, if an operation is waiting for a leader. https://github.com/ipfs/ipfs-cluster/blob/master/consensus/raft/raft.go#L280
